v2/yarpcgrpc/outbound.go

```diff
@@ -174,50 +177,48 @@ func metadataToPeer(responseMD metadata.MD) yarpc.Identifier {
 	return yarpc.Address(value[0])
 }
 
-func metadataToApplicationError(invokeErr error, responseMD metadata.MD) error {
-	if responseMD == nil {
-		return nil
-	}
-
-	if _, ok := responseMD[ApplicationErrorHeader]; ok {
-		return invokeErr
-	}
-	return nil
-}
-
-func invokeErrorToYARPCError(err error, responseMD metadata.MD) error {
+func invokeErrorToYARPCError(err error, responseMD metadata.MD) (*yarpcerror.Info, []byte, error) {
 	if err == nil {
-		return nil
+		return nil, nil, nil
 	}
-	if yarpcerror.IsStatus(err) {
-		return err
-	}
-	status, ok := status.FromError(err)
-	// if not a yarpc error or grpc error, just return a wrapped error
-	if !ok {
-		return yarpcerror.WrapError(err)
-	}
-	code, ok := _grpcCodeToCode[status.Code()]
+
+	s := status.Convert(err)
+	code, ok := _grpcCodeToCode[s.Code()]
 	if !ok {
 		code = yarpcerror.CodeUnknown
 	}
 	var name string
+	var isAppErr bool
 	if responseMD != nil {
 		value, ok := responseMD[ErrorNameHeader]
 		// TODO: what to do if the length is > 1?
 		if ok && len(value) == 1 {
 			name = value[0]
 		}
+
+		appErrHeader, ok := responseMD[ApplicationErrorHeader]
+		if ok && len(appErrHeader) == 1 {
+			isAppErr = appErrHeader[0] == ApplicationErrorHeaderValue
+		}
 	}
-	message := status.Message()
+	message := s.Message()
 	// we put the name as a prefix for grpc compatibility
 	// if there was no message, the message will be the name, so we leave it as the message
 	if name != "" && message != "" && message != name {
 		message = strings.TrimPrefix(message, name+": ")
 	} else if name != "" && message == name {
 		message = ""
 	}
-	return yarpcerror.New(code, message, yarpcerror.WithName(name))
+
+	if isAppErr {
+		buf, err := proto.Marshal(s.Proto())
+		if err != nil {
+			return nil, nil, err
```

Comment:
Should this last error be a yarpcerror?

Reply:
yarpcerrors are inspected using our GetInfo or GetExtract methods only, so calling a GetInfo on a non-yarpcerror will wrap it accordingly. In other words, we don't have to actually wrap any errors with a yarpcerror, unless we want to explicitly attach an error code or name to it.